subsys/net/lib/latmon/latmon.c

```diff
@@ -0,0 +1,458 @@
+			msg, conf, data,
+			MONITOR_THREAD_PRIORITY, 0, K_NO_WAIT);
+}
+
+static void stop_monitoring(void)
+{
+	if (!monitor_tid) {
+		return;
+	}
+
+	abort_monitor = true;
+	k_sem_take(&monitor_done, K_FOREVER);
+}
+
+static void handle_connection(struct latmon_message *msg)
+{
+#if (K_HEAP_MEM_POOL_SIZE > 0)
+	struct latmus_conf *conf = k_malloc(sizeof(*conf));
+	struct latmon_data *data = k_malloc(sizeof(*data));
+	struct latmon_net_request req;
+	ssize_t len;
+
+	if (!conf || !data) {
+		LOG_ERR("Failed to allocate memory, check HEAP_MEM_POOL_SIZE");
+		goto out;
+	}
+
+	memset(conf, 0, sizeof(*conf));
+	memset(data, 0, sizeof(*data));
+	for (;;) {
+		len = recv(msg->latmus, &req, sizeof(req), 0);
+		stop_monitoring();
+		if (get_latmus_conf(len, &req, conf) < 0) {
+			/* we might have to send the histogram */
+			if (send_trailing_data(msg->latmus, conf, data) < 0) {
+				break;
+			}
+			memset(conf, 0, sizeof(*conf));
+			continue;
+		}
+		start_monitoring(msg, conf, data);
+	}
+out:
+	k_free(conf);
+	k_free(data);
+	close(msg->latmus);
+	k_sem_give(&latmon_done);
+#else
+	LOG_ERR("No heap configured");
+#endif
+}
+
+static int latmon_server_thread(void *p1, void *p2, void *p3)
+{
+	struct latmon_message msg = { };
+
+	for (;;) {
+		k_msgq_get(&latmon_msgq, &msg, K_FOREVER);
+
+		/* Only latmus can stop the monitoring, so hang in there */
+		handle_connection(&msg);
+	}
+
+	return 0;
+}
+
+static void broadcast_ip_address(struct in_addr *ip_addr)
+{
+	char ip_str[NET_IPV4_ADDR_LEN];
+	struct sockaddr_in broadcast;
+	int sock;
+
+	sock = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
+	if (sock < 0) {
+		perror("socket");
+		exit(1);
+	}
+
+	broadcast.sin_addr.s_addr = htonl(INADDR_BROADCAST);
+	broadcast.sin_port = htons(LATMON_NET_PORT);
+	broadcast.sin_family = AF_INET;
+
+	net_addr_ntop(AF_INET, ip_addr, ip_str, sizeof(ip_str));
+
+	if (sendto(sock, ip_str, strlen(ip_str), 0,
+		(struct sockaddr *)&broadcast, sizeof(broadcast)) < 0) {
+		LOG_ERR("Failed to send broadcast message");
+	}
+
+	close(sock);
+}
+
+/* Get a socket to listen to Latmus requests */
+int net_latmon_get_socket(void)
+{
+	struct sockaddr_in bind_addr = {
+		.sin_family = AF_INET,
+		.sin_addr.s_addr = htonl(INADDR_ANY),
+		.sin_port = htons(LATMON_NET_PORT)
+	};
+	int s, on = 1;
+
+	s = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
+	if (s < 0) {
+		perror("socket");
+		exit(1);
+	}
+
+	setsockopt(s, SOL_SOCKET, SO_REUSEADDR, &on, sizeof(on));
+	if (bind(s, (struct sockaddr *)&bind_addr, sizeof(bind_addr)) < 0) {
+		perror("bind");
+		exit(1);
+	}
+
+	if (listen(s, 1) < 0) {
+		perror("listen");
+		exit(1);
+	}
+
+	return s;
+}
+
+/* Waits for connection from Latmus */
+int net_latmon_connect(int socket, struct in_addr *ip)
+{
+	struct pollfd fds[1] = { {.fd = socket, .events = POLLIN } };
+	struct sockaddr_in clnt_addr;
+	socklen_t len;
+	int latmus;
+	int ret;
+
+	LOG_INF("Waiting for connection...");
+
+	ret = poll(fds, 1, 5000);
```

Comment:
Please provide the timeout as a parameter to the `net_latmon_connect` API rather than a "magic" value, unless 5000 is defined by the protocol, in which case, please define it, something like `#define NET_LATMON_CONNECT_TIMEOUT 5000`